subprojects/core/src/integTest/groovy/org/gradle/api/internal/changedetection/CorruptedTaskHistoryIntegrationTest.groovy

```diff
@@ -0,0 +1,133 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
+
+class CorruptedTaskHistoryIntegrationTest extends AbstractIntegrationSpec {
+
+    // If this test starts to be flaky it points to a problem in the code!
+    // See the linked issue.
+    @Issue("https://github.com/gradle/gradle/issues/2827")
+    def "broken build doesn't corrupt the artifact history"() {
+        def numberOfFiles = 10
+        def numberOfInputProperties = 10
+        def numberOfTasks = 100
+        def millisecondsToKill = 300
+
+        setupTestProject(numberOfFiles, numberOfInputProperties, numberOfTasks, millisecondsToKill)
+
+        executer.beforeExecute {
+            // We need a separate JVM in order not to kill the test JVM
+            requireGradleDistribution()
+        }
+
+        when:
+        succeeds("createFiles")
+        succeeds("clean")
+        fails("createFiles", "-PkillMe=true", "--max-workers=${numberOfTasks}")
+
+        then:
+        file('build/output10').allDescendants().size() == numberOfFiles
+
+        expect:
+        succeeds "createFiles"
+    }
+
+    private void setupTestProject(int numberOfFiles, int numberOfInputProperties, int numberOfTasks, int millisecondsToKill) {
+        buildFile << """
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.workers.IsolationMode
+import org.gradle.workers.WorkerExecutor
+
+import javax.inject.Inject
+
+class CreateFiles implements Runnable {
+
+    private final File outputDir
+     
+    @Inject
+    CreateFiles(File outputDir) {
+        this.outputDir=outputDir
+    }
+
+    @Override
+    void run() {
+        (1..$numberOfFiles).each {
+            new File(outputDir, "\${it}.txt").text = "output\${it}"
+        }
+    }
+}            
+
+class CreateFilesTask extends DefaultTask {
+    private final WorkerExecutor workerExecutor
+    
+    @Inject
+    CreateFilesTask(WorkerExecutor workerExecutor) { 
+        this.workerExecutor=workerExecutor
+    }
+    
+    ${(1..numberOfInputProperties).collect { inputProperty(it) }.join("\n")}
+    
+    @OutputDirectory
+    File outputDir
+
+    @TaskAction
+    void createFiles() {
+        workerExecutor.submit(CreateFiles) {
+            isolationMode = IsolationMode.NONE
+            params(outputDir)
+        }
+        if (project.findProperty("killMe")) {
+            new Thread({
+                Thread.sleep(${millisecondsToKill})
+                System.exit(1)
+            }).start()
```

Comment:
Should we try to have a single kill thread/task?

Reply:
I wouldn't add the extra complexity. Like this we can kill the build some time after the first task finished writing its files. WDYT?